File: app/src/main/java/me/pipi/deliveries/data/TimelineSlot.java
package me.pipi.deliveries.data;

import java.util.Locale;

/** Canonical provider slots shared by cache writes and diagnostics; old names are read aliases. */
public final class TimelineSlot {
    public static final String V5_QUERY = "v5_query";
    public static final String V6_LIST = "v6_list";
    public static final String V4_QUERY = "v4_query";
    public static final String V6_QUERY = "v6_query";
    public static final String K100_H5 = "k100_h5";
    public static final String CN_H5 = "cn_h5";
    public static final String JD_H5 = "jd_h5";
    public static final String KDNIAO = "kdniao";

    private TimelineSlot() {}

    public static boolean isAutomaticH5(String value) {
        return CN_H5.equals(normalize(value)) || JD_H5.equals(normalize(value));
    }

    public static String normalize(String value) {
        String clean = value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
        switch (clean) {
            case "interface5":
                return V5_QUERY;
            case "interface6":
                return V6_LIST;
            case "v4":
                return V4_QUERY;
            case "meizu":
            case "meizu_picker":
            case "v6_picker":
                return V6_QUERY;
            case "kuaidi100":
            case "web":
                return K100_H5;
            default:
                return clean;
        }
    }

    /** 账号槽（v5_query / v6_list）：节点按账号接口的状态码解码。 */
    public static boolean isAccount(String value) {
        String slot = normalize(value);
        return V5_QUERY.equals(slot) || V6_LIST.equals(slot);
    }

    /** 账号槽对应的绑定来源；非账号槽返回空串。 */
    public static String bindingSourceOf(String value) {
        String slot = normalize(value);
        if (V5_QUERY.equals(slot)) return "interface5";
        if (V6_LIST.equals(slot)) return "interface6";
        return "";
    }

    /** 绑定来源对应的账号槽。 */
    public static String forBindingSource(String bindingSource) {
        String clean = bindingSource == null ? "" : bindingSource.trim().toLowerCase(Locale.ROOT);
        return "interface5".equals(clean) || V5_QUERY.equals(clean) ? V5_QUERY : V6_LIST;
    }
}

File: script/project/services/timeline-slot.ts
/**
 * Durable provider slots are independent whole-package caches. `v6_query` owns Meizu Online
 * history; legacy Meizu names normalize to this slot when reading stored rows.
 * `k100_h5` owns the fixed /app/query/?nu= page. `kuaidi100` is the unused paid poll slot.
 * Legacy names normalize once through normalizeTimelineSlot when reading stored rows.
 * `interface5` / `account` are feed packages (log level v5_list), not v5_query.
 */
export const TIMELINE_SLOT = {
  V5_QUERY: "v5_query",
  V4_QUERY: "v4_query",
  V6_QUERY: "v6_query",
  V2_QUERY: "v2_query",
  JD_H5: "jd_h5",
  CN_H5: "cn_h5",
  K100_H5: "k100_h5",
  KDNIAO: "kdniao",
  K100_PAID: "kuaidi100",
} as const;

export type TimelineSlot = typeof TIMELINE_SLOT[keyof typeof TIMELINE_SLOT];

/**
 * @param legacyWebSlot 旧的 `web` 包既可能是菜鸟 H5 也可能是 K100 页，读盘时按票据来源决定。
 */
export function normalizeTimelineSlot(
  value: unknown,
  legacyWebSlot: "cn_h5" | "k100_h5" = TIMELINE_SLOT.K100_H5,
): string {
  const clean = String(value || "").trim().toLowerCase();
  switch (clean) {
    case "local":
    case "moto":
      return TIMELINE_SLOT.V4_QUERY;
    case "route":
    case "meizu":
    case "meizu_picker":
    case "v6_picker":
      return TIMELINE_SLOT.V6_QUERY;
    case "oppo":
      return TIMELINE_SLOT.V2_QUERY;
    case "fallback":
      return TIMELINE_SLOT.KDNIAO;
    case "cainiao_h5":
      return TIMELINE_SLOT.CN_H5;
    case "kuaidi100_h5":
      return TIMELINE_SLOT.K100_H5;
    case "web":
      return legacyWebSlot;
    case "account_detail":
      return TIMELINE_SLOT.V5_QUERY;
    case "jingdong_h5":
      return TIMELINE_SLOT.JD_H5;
    default:
      return clean;
  }
}

export function isTimelineSlot(value: unknown, slot: string): boolean {
  return normalizeTimelineSlot(value) === slot;
}
